The ncclient log's XML is pretty printed

miott opened the pull request:
Fixes #7 

Added a log handler that trims hello message, pretty prints XML for session.

Regarding #6 , this is being fixed in pyATS Genie. Unfortunately, ncclient now adds the rpc-reply to the the DEBUG logging level instead of the INFO logging level.  I tried to stop that when I saw that pull request in ncclient but was ignored.  We will have to add the rpc-reply to that log after we pick it up from ncclient.

miott commented:
The ncclient logs are showing up in the pyats logs now, even without my proposed changes.  The problem I am solving is pretty print.

Python logging works in a hierarchy.   ``logger = logging.getLogger(__name__)`` creates a log named after the module.  ``nccl`` is the ncclient logger, is created in the module and the handler is added.  The Netconf class is a member of the module and a subclass of ncclient.manager which reports to the ncclient logger.  The end result is that all the logs from ncclient and logger show up in the pyats logs.

simingy commented:
@miott i know how logging works... but how does a handler at the child level affect the logging at top when it's propagated?

ohhh you're just the handler to modify the record object

ohh that's not a good idea, what if someone attach another copy of the handler?
you can technically have multiple copy of the handler.
these things are normally done by a formatter, but you're doing the fomatting in the handler to get by the fact that you don't own the formatter to whichever handler...

miott commented:
I did not think of doing this in a formatter.  Is that your suggestion?

simingy commented:
formatters are attached to handlers ... either way, you are not emitting something, you're actually cheating the system - a handler will get called `emit()` if its level are within setting level.
but you are not emitting, you just rely on the mechanism to change LogRecord object.

but that also changes the truthfulness of the record - it's modified for everyone upstream

miott commented:
We have used this reformatting of ncclient logging for years without issue so this is not a new concept.

The basic issue I am trying to solve is the fact that the NETCONF messages are very hard to read if they are not pretty printed and ncclient is the one who constructs the log message.  We are relying on their poorly formatted logger.  I can remove the trimming down of the hello message if that is a worry.

We are the endpoint user so I am not sure we need to be concerned about "everyone upstream".

Let me know if this is a showstopper change.  I will have to wrangle with the ncclient team and figure out how to fix this in their library if I can't fix it in this one.

bastell commented:
The formatting change makes sense, as long as no information is discarded. The hello messages should be left as they are, since trimming them modifies the truthfulness of the logs. Additionally, there should be handling for bad xml. If the handler fails to load or write xml for whatever reason, it should leave the message alone instead of raising an exception.

miott commented:
Addressed review comments
```
2020-06-02T08:49:42: %NCCLIENT-INFO: Connected (version 2.0, client OpenSSH_6.1)
2020-06-02T08:49:43: %NCCLIENT-INFO: Authentication (publickey) failed.
2020-06-02T08:49:43: %NCCLIENT-INFO: Authentication (password) successful!
2020-06-02T08:49:43: %NCCLIENT-INFO: [host 172.27.255.22 session 0x111a2c050] Sending:
2020-06-02T08:49:43: %NCCLIENT-INFO: <nc:hello xmlns:nc="urn:ietf:params:xml:ns:netconf:base:1.0">
2020-06-02T08:49:43: %NCCLIENT-INFO:   <nc:capabilities>
2020-06-02T08:49:43: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:base:1.0</nc:capability>
2020-06-02T08:49:43: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:base:1.1</nc:capability>
2020-06-02T08:49:43: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:writable-running:1.0</nc:capability>
2020-06-02T08:49:43: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:candidate:1.0</nc:capability>
2020-06-02T08:49:43: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:confirmed-commit:1.0</nc:capability>
2020-06-02T08:49:43: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:rollback-on-error:1.0</nc:capability>
2020-06-02T08:49:43: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:startup:1.0</nc:capability>
2020-06-02T08:49:43: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:url:1.0?scheme=http,ftp,file,https,sftp</nc:capability>
2020-06-02T08:49:43: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:validate:1.0</nc:capability>
2020-06-02T08:49:43: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:xpath:1.0</nc:capability>
2020-06-02T08:49:43: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:notification:1.0</nc:capability>
2020-06-02T08:49:43: %NCCLIENT-INFO:     <nc:capability>urn:liberouter:params:netconf:capability:power-control:1.0</nc:capability>
2020-06-02T08:49:43: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:interleave:1.0</nc:capability>
2020-06-02T08:49:43: %NCCLIENT-INFO:     <nc:capability>urn:ietf:params:netconf:capability:with-defaults:1.0</nc:capability>
2020-06-02T08:49:43: %NCCLIENT-INFO:   </nc:capabilities>
2020-06-02T08:49:43: %NCCLIENT-INFO: </nc:hello>
2020-06-02T08:49:43: %NCCLIENT-INFO: ]]>]]>
2020-06-02T08:49:43: %NCCLIENT-INFO: [host 172.27.255.22 session 0x111a2c050] Received message from host
2020-06-02T08:49:43: %NCCLIENT-INFO: [host 172.27.255.22 session-id 528] initialized: session-id=528 | server_capabilities=<dict_keyiterator object at 0x111b47350>
2020-06-02T08:49:43: %YANG-INFO: +------------------------------------------------------------------------------+
2020-06-02T08:49:43: %YANG-INFO: |                              NETCONF CONNECTED                               |
2020-06-02T08:49:43: %YANG-INFO: +------------------------------------------------------------------------------+
```

bastell commented on connector/src/yang/connector/netconf.py:
```diff
@@ -39,6 +40,57 @@ class BaseConnection:
+                        record.args[i] = (arg[:start] +
+                                          text +
+                                          arg[end:]).decode()
+                    except:
```
```python
except (et.Error, UnicodeError):
```
So that BaseExceptions aren't suppressed. Otherwise looks good.

miott replied:
Only thing I can get bad parameters to throw are either TypeError for non-element values sent to et.tostring or AttributeError if .decode() is not there.  Really, TypeError would be the only one to catch that would be relevant to et.tostring.  Why do you suggest et.Error and UnicodeError?

miott replied:
Come to think of it, if my new formatting code fails to parse the record for whatever reason, I want _all_ of those errors to be suppressed and allow the record to continue as-is.  I should expand the try: except:.
